### PR TITLE
Add ripple adder construction and transpilation benchmarks.

### PR DESCRIPTION
**Pull request by kdk**
### Summary

Adds benchmarks based on Terra's examples/python/rippleadd.py :

```
ripple_adder.RippleAdderConstruction.peakmem_build_ripple_adder
ripple_adder.RippleAdderConstruction.time_build_ripple_adder
ripple_adder.RippleAdderTranspile.peakmem_transpile_simulator_ripple_adder
ripple_adder.RippleAdderTranspile.peakmem_transpile_square_grid_ripple_adder
ripple_adder.RippleAdderTranspile.time_transpile_simulator_ripple_adder
ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder
ripple_adder.RippleAdderTranspile.depth_transpile_simulator_ripple_adder
ripple_adder.RippleAdderTranspile.depth_transpile_square_grid_ripple_adder
```

### CODE REVIEW

File: test/benchmarks/ripple_adder.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 IBM RESEARCH. All Rights Reserved.
```
Comment (mtreinish) [inline test/benchmarks/ripple_adder.py]: I think this needs to be updated to match the rest of the project

**Review by mtreinish — COMMENTED** [review]
Do you have a sense of run time? I realize this will be pretty fast after all the recent terra performance work the past couple of weeks. I am asking because we're adding >80 benchmark permutations here and we're hitting the limits of what we can run in 24 hours. For example, yesterday the server took 22hrs 39min to finish running benchmarks on the 7 commits merged the day prior.

File: test/benchmarks/ripple_adder.py (hunk 1 above)
Comment (kdk) [reply test/benchmarks/ripple_adder.py]: Fixed in b4b1d23 .

**Comment by kdk** [comment]
Here's the timing info I saw locally running from https://github.com/Qiskit/qiskit-terra/commit/05355b103817dcd7aae0839a485b7ce28fb354a5 :

```
$ asv show --bench ripple master
Commit: 05355b10 <master>

ripple_adder.RippleAdderConstruction.peakmem_build_ripple_adder
...
duration: 17.7s

ripple_adder.RippleAdderConstruction.time_build_ripple_adder
...
duration: 36.4s

ripple_adder.RippleAdderTranspile.peakmem_transpile_simulator_ripple_adder
...
duration: 35.2s

ripple_adder.RippleAdderTranspile.peakmem_transpile_square_grid_ripple_adder
...
duration: 5.08m

ripple_adder.RippleAdderTranspile.time_transpile_simulator_ripple_adder
...
duration: 1.41m

ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder
...
duration: 9.32m

ripple_adder.RippleAdderTranspile.track_depth_transpile_simulator_ripple_adder
...
duration: 36.0s

ripple_adder.RippleAdderTranspile.track_depth_transpile_square_grid_ripple_adder 
...
duration: 5.01m
```

**Comment by mtreinish** [comment]
That concerns me a bit because we're adding 22minutes per commit, so for a day like Tuesday we landed seven commits that's an extra 2.5 hours of run time. Which would put it over 24hours. It's not the end of the world but it means we're pushing back the next days commits' benchmarks 24 hours if we miss the cron job. Maybe, we should reduce the number permutations a bit.

I'd like to say retworkx and hash caching will improve things here, but in my experience most of the run time in asv seems to be setup and not the actual benchmarks. So I'm not sure it'll make that much of a difference.

I'm thinking that I might need to rework the benchmarking script so that we remove the 24 hour trigger once restriction and rework it to queue up a run with a more frequent cron job that checks if we blocked a run. So we don't end up wasting most of a day sometimes.

**Comment by mtreinish** [comment]
ok, with #845 merging I'm fine with this as is. That should open up more than enough of a slot for these benchmarks.

**Review by mtreinish — APPROVED** [review]

**Comment by kdk** [comment]
> That concerns me a bit because we're adding 22minutes per commit, so for a day like Tuesday we landed seven commits that's an extra 2.5 hours of run time. Which would put it over 24hours. It's not the end of the world but it means we're pushing back the next days commits' benchmarks 24 hours if we miss the cron job. Maybe, we should reduce the number permutations a bit.

Fair point, I removed the `size=50` param for the transpiler benchmarks in 5308c26  , which takes the timing to:
```
$ asv show --bench ripple 05355b10 | grep -e local -e duration
ripple_adder.RippleAdderConstruction.peakmem_build_ripple_adder
  started: 2020-03-13 10:59:41, duration: 19.2s
ripple_adder.RippleAdderConstruction.time_build_ripple_adder
  started: 2020-03-13 11:00:00, duration: 39.5s
ripple_adder.RippleAdderTranspile.peakmem_transpile_simulator_ripple_adder
  started: 2020-03-13 11:00:40, duration: 19.3s
ripple_adder.RippleAdderTranspile.peakmem_transpile_square_grid_ripple_adder
  started: 2020-03-13 11:01:00, duration: 1.19m
ripple_adder.RippleAdderTranspile.time_transpile_simulator_ripple_adder
  started: 2020-03-13 11:02:11, duration: 48.8s
ripple_adder.RippleAdderTranspile.time_transpile_square_grid_ripple_adder
  started: 2020-03-13 11:03:00, duration: 2.45m
ripple_adder.RippleAdderTranspile.track_depth_transpile_simulator_ripple_adder
  started: 2020-03-13 11:05:27, duration: 19.5s
ripple_adder.RippleAdderTranspile.track_depth_transpile_square_grid_ripple_adder
  started: 2020-03-13 11:05:47, duration: 1.20m
```

**Pull request merged**